qa/tasks/vault.py

```diff
@@ -0,0 +1,231 @@
+"""
+Deploy and configure Vault for Teuthology
+"""
+
+import argparse
+import contextlib
+import logging
+import time
+
+import httplib
+import json
+
+from teuthology import misc as teuthology
+from teuthology import contextutil
+from teuthology.orchestra import run
+from teuthology.exceptions import ConfigError
+
+
+log = logging.getLogger(__name__)
+
+
+def assign_ports(ctx, config, initial_port):
+    """
+    Assign port numbers starting from @initial_port
+    """
+    port = initial_port
+    role_endpoints = {}
+    for remote, roles_for_host in ctx.cluster.remotes.iteritems():
+        for role in roles_for_host:
+            if role in config:
+                role_endpoints[role] = (remote.name.split('@')[1], port)
+                port += 1
+
+    return role_endpoints
+
+
+@contextlib.contextmanager
+def download(ctx, config):
+    """
+    Download Vault Release from Hashicopr website.
+    Remove downloaded file upon exit.
+    """
+    assert isinstance(config, dict)
+    log.info('Downloading Vault...')
+    testdir = teuthology.get_testdir(ctx)
+
+    for (client, cconf) in config.items():
+        vault_version = cconf.get('version', '1.2.2')
+
+        ctx.cluster.only(client).run(
+            args=['mkdir', '-p', '{tdir}/vault'.format(tdir=testdir)])
+
+        cmd = [
+            'curl', '-L',
+            'https://releases.hashicorp.com/vault/{version}/vault_{version}_linux_amd64.zip'.format(version=vault_version), '-o',
```

Comment:
Oh my.  Ugh.  This isn't really your fault, the qa code has this pattern all over.  But it is a problem; you've got something here that downloads stuff from some external source.  And, um, 1.2.3 is out now.  And the integrity check you're doing isn't ideal.

So the kinds of problems this creates: somebody doing an ultra-secure code build (on an isolated network) can't run the test code because it has all these wired-in DNS dependencies.  Or, any of us, trying to build/test ceph at an airport with broken internet.  A code archeologist 20 years from now also won't be able to build this (ok maybe we don't care?).

I don't think you can fix all of this here.  But I'd like to suggest that, in the interest of future sanity, you do: /1/ checksum the file you download and compare it to a saved hash, and /2/ the URL you're downloading, the version you're using, and the checksum should be in named constants (say at the start of this file) and not embedded in the code.  I guess you've got logic here already to override the version - so maybe you want something different.  This really needs a much more global solution, and there's all sorts of other hassle making it all work nicely.

Reply:
You're absolutely right, this is crazy but sadly seems to be the norm in teuthology and we didn't know how to change it. I think your suggestions for adding a checksum and URL parameters make sense and we'll implement them in the next iteration if that's okay?